matchbox_socket/src/webrtc_socket/native.rs

```diff
@@ -82,127 +83,222 @@ impl Signaller for NativeSignaller {
+    );
+
+    async fn offer_handshake(
+        signal_peer: SignalPeer,
+        mut from_peer_rx: UnboundedReceiver<PeerSignal>,
+        peer_state_tx: UnboundedSender<(PeerId, PeerState)>,
+        messages_from_peers_tx: Vec<UnboundedSender<(PeerId, Packet)>>,
+        config: &WebRtcSocketConfig,
+    ) -> (PeerId, Vec<Self::DataChannel>, Self::PeerLoopArgs) {
+        let (to_peer_message_tx, to_peer_message_rx) = new_senders_and_receivers(config);
+
+        debug!("making offer");
+        let (connection, trickle) =
+            create_rtc_peer_connection(signal_peer.clone(), peer_state_tx.clone(), config)
+                .compat()
+                .await
+                .unwrap();
+
+        let (channel_ready_tx, wait_for_channels) = create_data_channels_ready_fut(config);
+        let data_channels = create_data_channels(
+            &connection,
+            channel_ready_tx,
+            signal_peer.id.clone(),
+            peer_state_tx.clone(),
+            messages_from_peers_tx,
+            &config.channels,
+        )
+        .await;
+
+        // TODO: maybe pass in options? ice restart etc.?
+        let offer = connection.create_offer(None).compat().await.unwrap();
+        let sdp = offer.sdp.clone();
+        connection
+            .set_local_description(offer)
             .compat()
             .await
-    }
-}
+            .unwrap();
+        signal_peer.send(PeerSignal::Offer(sdp));
 
-async fn message_loop_impl(
-    mut id_tx: Sender<PeerId>,
-    config: &WebRtcSocketConfig,
-    channels: MessageLoopChannels,
-) {
-    let MessageLoopChannels {
-        requests_sender,
-        mut events_receiver,
-        mut peer_messages_out_rx,
-        peer_state_tx,
-        messages_from_peers_tx,
-    } = channels;
-    debug!("Entering native WebRtcSocket message loop");
-
-    let mut peer_loops = FuturesUnordered::new();
-    let mut handshake_signals = HashMap::new();
-    let mut connected_peers = HashMap::new();
-
-    let timeout = Delay::new(Duration::from_millis(KEEP_ALIVE_INTERVAL));
-    futures::pin_mut!(timeout);
+        let answer = loop {
+            let signal = from_peer_rx
+                .next()
+                .await
+                .expect("Signal server connection lost in the middle of a handshake");
 
-    loop {
-        let mut next_peer_messages_out: FuturesUnordered<_> = peer_messages_out_rx
-            .iter_mut()
-            .enumerate()
-            .map(|(index, rx)| async move { (index, rx.next().await) })
-            .collect();
+            match signal {
+                PeerSignal::Answer(answer) => {
+                    break answer;
+                }
+                PeerSignal::Offer(_) => {
+                    warn!("Got an unexpected Offer, while waiting for Answer. Ignoring.")
+                }
+                PeerSignal::IceCandidate(_) => {
+                    warn!("Got an unexpected IceCandidate, while waiting for Answer. Ignoring.")
+                }
```

Comment:
in accept_handshake, you don't perform an exhaustive match, unlike here:
```
                _ => {
                    warn!("ignoring other signal!!!");
                }
```

I don't care which you choose, but consistency would be better

Reply:
Didn't actually touch this bit of code, it's copied from the previous implementation - as are the others

Reply:
OK - good with me then.

Reply:
I shall, but will leave it to a future PR